diff --git a/hw/block/virtio-blk.c b/hw/block/virtio-blk.c
index bac2d6fa2..98e0b5502 100644
--- a/hw/block/virtio-blk.c
+++ b/hw/block/virtio-blk.c
@@ -26,6 +26,8 @@
 #include "hw/virtio/virtio-blk.h"
 #include "dataplane/virtio-blk.h"
 #include "scsi/constants.h"
+
+extern int wavevm_blk_interceptor(uint64_t sector, QEMUIOVector *qiov, int is_write);
 #ifdef __linux__
 # include <scsi/sg.h>
 #endif
@@ -672,6 +674,12 @@ static int virtio_blk_handle_request(VirtIOBlockReq *req, MultiReqBuffer *mrb)
                                          req->qiov.size / BDRV_SECTOR_SIZE);
         }
 
+        if (wavevm_blk_interceptor(req->sector_num, &req->qiov, is_write) == 0) {
+            virtio_blk_req_complete(req, VIRTIO_BLK_S_OK);
+            virtio_blk_free_request(req);
+            return 0;
+        }
+
         if (!virtio_blk_sect_range_ok(s, req->sector_num, req->qiov.size)) {
             virtio_blk_req_complete(req, VIRTIO_BLK_S_IOERR);
             block_acct_invalid(blk_get_stats(s->blk),
